src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs

```diff
@@ -0,0 +1,180 @@
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+
+            try
+            {
+                testPlatformEventSource.MultiTestRunFinalizationStart(attachments?.Count ?? 0);
+                requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAttachmentsSentForFinalization, attachments?.Count ?? 0);
+                
+                cancellationToken.ThrowIfCancellationRequested();                
+
+                var taskCompletionSource = new TaskCompletionSource<Collection<AttachmentSet>>();
+                using (cancellationToken.Register(() => taskCompletionSource.TrySetCanceled()))
+                {
+                    Task<Collection<AttachmentSet>> task = Task.Run(async () => await ProcessAttachmentsAsync(new Collection<AttachmentSet>(attachments.ToList()), eventHandler, cancellationToken));
+
+                    var completedTask = await Task.WhenAny(task, taskCompletionSource.Task).ConfigureAwait(false);
+
+                    if (completedTask == task)
+                    {
+                        return FinalizeOperation(requestData, new MultiTestRunFinalizationCompleteEventArgs(false, null), await task, stopwatch, eventHandler);
+                    }
+                    else
+                    {
+                        eventHandler?.HandleLogMessage(TestMessageLevel.Informational, "Finalization was cancelled.");
+                        return FinalizeOperation(requestData, new MultiTestRunFinalizationCompleteEventArgs(true, null), attachments, stopwatch, eventHandler);
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                if (EqtTrace.IsWarningEnabled)
+                {
+                    EqtTrace.Warning("MultiTestRunFinalizationManager: operation was cancelled.");
+                }
+                return FinalizeOperation(requestData, new MultiTestRunFinalizationCompleteEventArgs(true, null), attachments, stopwatch, eventHandler);
+            }
+            catch (Exception e)
+            {
+                EqtTrace.Error("MultiTestRunFinalizationManager: Exception in FinalizeMultiTestRunAsync: " + e);
+
+                eventHandler?.HandleLogMessage(TestMessageLevel.Error, e.Message);
+                return FinalizeOperation(requestData, new MultiTestRunFinalizationCompleteEventArgs(false, e), attachments, stopwatch, eventHandler);
+            }
+        }
+
+        private async Task<Collection<AttachmentSet>> ProcessAttachmentsAsync(Collection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventsHandler, CancellationToken cancellationToken)
+        {
+            if (attachments == null || !attachments.Any()) return attachments;
+
+            var logger = CreateMessageLogger(eventsHandler);
+
+            for (int i = 0; i < dataCollectorAttachmentsProcessors.Length; i++)
```

Comment:
>dataCollectorAttachmentsProcessors.Length [](start = 32, length = 41)

I'm guessing the ordering of attachment processors is a TODO in this PR? Perhaps we could add a note and bring that in another PR.

Reply:
Do we want to expose priority on Attachment Processor interface?

Reply:
We can add it later as attribute. Resolving